rgw/curl: async curl client

cbodley opened the pull request:
adds class `rgw::curl::Client` that manages a curl multi handle and its associated connection pool. member function `async_perform()` executes an asynchronous http request as if by calling `curl_easy_perform()`. this function is not thread-safe, and must be called from the same executor that the Client was constructed with

also adds a free function `rgw::curl::async_perform()` that is thread-safe, and uses a hidden 'global' instance of Client that's wrapped in a strand executor. that global Client's lifetime is associated with the `asio::io_context` of its executor

the `Client` class is well-suited to single-threaded use cases like multisite sync. the free function is better suited to general use where you want to share a connection pool across several threads

```c++
// send a GET request for the resource specified by url
// errors are thrown as boost::system::system_error exceptions
void get_resource(const char* url, boost::asio::yield_context yield)
{
  auto easy = rgw::curl::easy_init();
  
  curl_easy_setopt(easy.get(), CURLOPT_URL, url);
  // set other options like CURLOPT_HEADERFUNCTION, CURLOPT_WRITEFUNCTION etc
  
  rgw::curl::async_perform(yield.get_executor(), easy.get(), yield);
}
```

---

unlike the existing `RGWHTTPManager` based on curl multi, `rgw::curl::Client` uses the 'multi-socket' api which avoids the need to dedicate a background thread to polling for curl completions. instead, it manages its own sockets and timer with asio that run in an existing event loop (asio::io_context):
```c++
// This implementation uses the 'multi_socket' flavor of the libcurl multi API:
// https://curl.se/libcurl/c/libcurl-multi.html
//
// By providing our own CURLMOPT_TIMERFUNCTION AND CURLMOPT_SOCKETFUNCTION, we
// can do all necessary polling and waiting on the given asio executor. As we
// get wakeups from asio, we call curl_multi_socket_action() to do its
// non-blocking socket i/o from within that executor.
//
// This also overrides CURLOPT_OPENSOCKETFUNCTION on each request to manage the
// pool of asio sockets.
```

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

cbodley committed:
rgw/curl: async curl client

adds class rgw::curl::Client that manages a curl multi handle and its
associated connection pool. member function async_perform() executes an
asynchronous http request as if by calling curl_easy_perform(). this
function is not thread-safe, and must be called from the same executor
that the Client was constructed with

unlike the existing RGWHTTPManager based on curl multi, rgw::curl::Client
uses the 'multi-socket' api which avoids the need to dedicate a background
thread to polling for curl completions. instead, it manages its own sockets
and timer with asio that run in an existing event loop (asio::io_context)

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/curl: switch to generic protocol sockets

unlike the concrete boost::asio::ip::tcp::socket and udp::socket types,
we can map the opensocket_callback's curl_sockaddr argument directly to
these generic protocol socket constructors

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/curl: add async_perform() free function

adds a free function rgw::curl::async_perform() that is thread-safe,
and uses a hidden 'global' instance of Client that's wrapped in a
strand executor. that global Client's lifetime is associated with the
asio::io_context of its executor

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented:
@adamemerson you had included most of this in https://github.com/ceph/ceph/pull/54535, but the thread-safe free function is new

cbodley commented:
jenkins test api

cbodley commented:
jenkins test make check

cbodley commented:
no changes, just pushed a rebase

mheler committed:
rgw/curl: fix libcurl socket callbacks

libcurl can pass resolver-owned sockets to the multi socket callback.
Those sockets are not in our socket map, so wait on them without taking
ownership.

Also mask SOCK_CLOEXEC and SOCK_NONBLOCK before matching the socket type
in opensocket_callback, because libcurl may include those flags in
socktype.

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>

mheler committed:
rgw/curl: log libcurl verbose output

When RGW debug logging is verbose, enable CURLOPT_VERBOSE and route
libcurl's text trace through the RGW log.

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>

RaminNietzsche commented:
@cbodley @adamemerson Rebased `wip-rgw-async-curl` onto current `main` (`93c3f6703157`) and pushed to my fork (`RaminNietzsche/ceph`). Rebase was clean (no conflicts).

**Local verification**
- `make` completed successfully
- `ninja unittest_rgw_curl_client` + `./bin/unittest_rgw_curl_client` — **13/13 PASSED** (~441 ms)

**Fork branch:** https://github.com/RaminNietzsche/ceph/tree/wip-rgw-async-curl (`4691db510d2`)

**Commit series on my fork (oldest → newest)**
```
20bd50e rgw/curl: async curl client
a1f69de rgw/curl: switch to generic protocol sockets
4691db5 rgw/curl: add async_perform() free function
```

The PR head is still `cbodley/ceph:wip-rgw-async-curl` at `81485919ab1`. To refresh the PR branch onto current `main`:

```bash
git fetch https://github.com/RaminNietzsche/ceph.git wip-rgw-async-curl
git checkout wip-rgw-async-curl
# or cherry-pick the series onto your branch
```

Nov 2025 CI failures (`make check`, arm64, api) may be stale after rebase — happy to help investigate if a re-run would be useful.

cbodley committed:
test/rgw: start testing for name resolution

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/curl: adjust masking of SOCK_CLOEXEC and SOCK_NONBLOCK

cbodley committed:
cmake/rgw: link rgw_curl as separate library target

makes it easier to apply BOOST_ASIO_ENABLE_HANDLER_TRACKING to the unit
test without recompiling all of rgw_common

also link global and unit-main so that --debug-rgw can be used to enable
libcurl verbose logging

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented:
i included two of @mheler's commits from https://github.com/ceph/ceph-ci/commits/wip-restore-async-curl/ and added unit tests to exercise the flows with curl-managed sockets for name resolution. the `Client.resolve_http` test succeeds, but `Client.resolve_https` hangs after sending the request. still trying to debug that, but it doesn't look like libcurl ever asks us to wait for the response

cbodley commented:
some logs captured from `bin/unittest_rgw_curl_client --debug-rgw 20` with `BOOST_ASIO_ENABLE_HANDLER_TRACKING` and some additional debug logging around socket_callback and socket_action:

<details>

<summary>output from Client.resolve_https</summary>

```
[ RUN      ] Client.resolve_https
@asio|1783969890.036647|0^1|in 'co_spawn_entry_point' (/home/cbodley/ceph/build/boost/include/boost/asio/impl/co_spawn.hpp:197)
@asio|1783969890.036647|0*1|io_context@0x7ffdd46a7980.execute
@asio|1783969890.036692|>1|
@asio|1783969890.036721|1*2|deadline_timer@0x55cfc18cd660.async_wait
> socket action on fd=-1 events=0
socket callback fd=11 what=1
@asio|1783969890.036965|1*3|descriptor@0x55cfc1a3e888.async_wait
@asio|1783969890.036971|1|deadline_timer@0x55cfc18cd660.cancel
@asio|1783969890.036986|1*4|deadline_timer@0x55cfc18cd660.async_wait
< socket action on -1 ok
@asio|1783969890.036995|<1|
@asio|1783969890.036998|>2|ec=system:125
@asio|1783969890.037001|<2|
@asio|1783969890.038496|>4|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969890.038548|4|deadline_timer@0x55cfc18cd660.cancel
@asio|1783969890.038569|4*5|deadline_timer@0x55cfc18cd660.async_wait
< socket action on -1 ok
@asio|1783969890.038577|<4|
@asio|1783969890.041044|>5|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969890.041082|5|deadline_timer@0x55cfc18cd660.cancel
@asio|1783969890.041104|5*6|deadline_timer@0x55cfc18cd660.async_wait
< socket action on -1 ok
@asio|1783969890.041113|<5|
@asio|1783969890.048188|>6|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969890.048254|6|deadline_timer@0x55cfc18cd660.cancel
@asio|1783969890.048266|6*7|deadline_timer@0x55cfc18cd660.async_wait
< socket action on -1 ok
@asio|1783969890.048274|<6|
@asio|1783969890.056831|>7|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969890.056892|7|deadline_timer@0x55cfc18cd660.cancel
@asio|1783969890.056899|7*8|deadline_timer@0x55cfc18cd660.async_wait
< socket action on -1 ok
@asio|1783969890.056904|<7|
@asio|1783969890.065784|>3|ec=system:0
> socket action on fd=11 events=1
socket callback fd=11 what=4
@asio|1783969890.065851|3|descriptor@0x55cfc1a3e888.release
Host ceph.io:443 was resolved.
IPv6: 2600:1f18:532a:5b0a:d814:de05:a120:8445
IPv4: 100.30.106.20
  Trying [2600:1f18:532a:5b0a:d814:de05:a120:8445]:443...
Immediate connect fail for 2600:1f18:532a:5b0a:d814:de05:a120:8445: Network is unreachable
@asio|1783969890.066219|3|socket@0x55cfc1a3e888.close
  Trying 100.30.106.20:443...
socket callback fd=12 what=2
@asio|1783969890.066386|3*9|socket@0x55cfc1a3e888.async_wait
< socket action on 11 ok
@asio|1783969890.066394|<3|
@asio|1783969890.073242|>8|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969890.073383|8|deadline_timer@0x55cfc18cd660.cancel
< socket action on -1 ok
@asio|1783969890.073406|<8|
@asio|1783969890.085432|>9|ec=system:0
> socket action on fd=12 events=2
ALPN: curl offers h2,http/1.1
TLSv1.3 (OUT), TLS handshake, Client hello (1):
SSL Trust Anchors:
  CAfile: /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem
socket callback fd=12 what=1
@asio|1783969890.092383|9*10|socket@0x55cfc1a3e888.async_wait
< socket action on 12 ok
@asio|1783969890.092395|<9|
@asio|1783969890.174879|>10|ec=system:0
> socket action on fd=12 events=1
TLSv1.3 (IN), TLS handshake, Server hello (2):
TLSv1.3 (IN), TLS change cipher, Change cipher spec (1):
TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
TLSv1.3 (IN), TLS handshake, Certificate (11):
TLSv1.3 (IN), TLS handshake, CERT verify (15):
TLSv1.3 (IN), TLS handshake, Finished (20):
TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
TLSv1.3 (OUT), TLS handshake, Finished (20):
SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384 / x25519 / RSASSA-PSS
ALPN: server accepted http/1.1
Server certificate:
  subject: CN=ceph.io
  start date: May  3 14:48:36 2026 GMT
  expire date: Aug  1 14:48:35 2026 GMT
  issuer: C=US; O=Let's Encrypt; CN=R13
  Certificate level 0: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
  Certificate level 1: Public key type RSA (2048/112 Bits/secBits), signed using sha256WithRSAEncryption
  Certificate level 2: Public key type RSA (4096/152 Bits/secBits), signed using sha256WithRSAEncryption
  subjectAltName: "ceph.io" matches cert's "ceph.io"
SSL certificate verified via OpenSSL.
Established connection to ceph.io (100.30.106.20 port 443) from 192.168.245.130 port 59326
using HTTP/1.x
Request completely sent off
< socket action on 12 ok
@asio|1783969890.176049|<10|
^C
```

</details>

<details>

<summary>same flow from Client.resolve_http</summary>

```
[ RUN      ] Client.resolve_http
@asio|1783969807.971209|0^1|in 'co_spawn_entry_point' (/home/cbodley/ceph/build/boost/include/boost/asio/impl/co_spawn.hpp:197)
@asio|1783969807.971209|0*1|io_context@0x7ffcd4255dc0.execute
@asio|1783969807.971228|>1|
@asio|1783969807.971252|1*2|deadline_timer@0x55a2c5be4660.async_wait
> socket action on fd=-1 events=0
socket callback fd=11 what=1
@asio|1783969807.971675|1*3|descriptor@0x55a2c5d55888.async_wait
@asio|1783969807.971705|1|deadline_timer@0x55a2c5be4660.cancel
@asio|1783969807.971709|1*4|deadline_timer@0x55a2c5be4660.async_wait
< socket action on -1 ok
@asio|1783969807.971717|<1|
@asio|1783969807.971721|>2|ec=system:125
@asio|1783969807.971786|<2|
@asio|1783969807.972893|>4|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969807.972924|4|deadline_timer@0x55a2c5be4660.cancel
@asio|1783969807.972928|4*5|deadline_timer@0x55a2c5be4660.async_wait
< socket action on -1 ok
@asio|1783969807.972948|<4|
@asio|1783969807.974964|>5|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969807.974999|5|deadline_timer@0x55a2c5be4660.cancel
@asio|1783969807.975016|5*6|deadline_timer@0x55a2c5be4660.async_wait
< socket action on -1 ok
@asio|1783969807.975023|<5|
@asio|1783969807.982347|>6|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969807.982457|6|deadline_timer@0x55a2c5be4660.cancel
@asio|1783969807.982465|6*7|deadline_timer@0x55a2c5be4660.async_wait
< socket action on -1 ok
@asio|1783969807.982484|<6|
@asio|1783969807.990499|>7|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969807.990563|7|deadline_timer@0x55a2c5be4660.cancel
@asio|1783969807.990584|7*8|deadline_timer@0x55a2c5be4660.async_wait
< socket action on -1 ok
@asio|1783969807.990591|<7|
@asio|1783969808.001258|>3|ec=system:0
> socket action on fd=11 events=1
socket callback fd=11 what=4
@asio|1783969808.001322|3|descriptor@0x55a2c5d55888.release
Host ceph.io:80 was resolved.
IPv6: 2600:1f18:532a:5b0a:d814:de05:a120:8445
IPv4: 100.30.106.20
  Trying [2600:1f18:532a:5b0a:d814:de05:a120:8445]:80...
Immediate connect fail for 2600:1f18:532a:5b0a:d814:de05:a120:8445: Network is unreachable
@asio|1783969808.001431|3|socket@0x55a2c5d55888.close
  Trying 100.30.106.20:80...
socket callback fd=12 what=2
@asio|1783969808.001577|3*9|socket@0x55a2c5d55888.async_wait
< socket action on 11 ok
@asio|1783969808.001584|<3|
@asio|1783969808.006657|>8|ec=system:0
> socket action on fd=-1 events=0
@asio|1783969808.006801|8|deadline_timer@0x55a2c5be4660.cancel
< socket action on -1 ok
@asio|1783969808.006814|<8|
@asio|1783969808.018539|>9|ec=system:0
> socket action on fd=12 events=2
Established connection to ceph.io (100.30.106.20 port 80) from 192.168.245.130 port 56698
using HTTP/1.x
Request completely sent off
socket callback fd=12 what=1
@asio|1783969808.018852|9*10|socket@0x55a2c5d55888.async_wait
< socket action on 12 ok
@asio|1783969808.018883|<9|
@asio|1783969808.087876|>10|ec=system:0
> socket action on fd=12 events=1
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.18.0 (Ubuntu)</center>
</body>
</html>
Connection #0 to host ceph.io:80 left intact
socket callback fd=12 what=4
< socket action on 12 ok
socket status on 12 completing
@asio|1783969808.088084|<10|
@asio|1783969808.088118|0|socket@0x55a2c5d55888.close
[       OK ] Client.resolve_http (118 ms)
```

</details>

edit: some of the logs were out order because curl's output went through dout to a separate thread. i changed the debug_callback() to std::cout.write(data, size); so all the output goes to the same channel

cbodley committed:
rgw/curl: inline ext_socket_action() in socket_callback()

socket_callback() handles the two cases (client-managed vs curl-managed
sockets) in a somewhat-uniform way

reuse socket_wait_handler's call operator instead of calling
socket.async_wait() directly, taking advantage of the symmetry between
std::visit() and std::invoke()

cbodley committed:
rgw/curl: avoid posix dependency with generic stream socket

while not strictly necessary for use in rgw, the goal for this library
was to be as platform-independent as its curl and asio dependencies

cbodley commented:
i added a couple of cleanup commits, but they don't resolve the test failure above

mheler commented:
> i included two of @mheler's commits from https://github.com/ceph/ceph-ci/commits/wip-restore-async-curl/ and added unit tests to exercise the flows with curl-managed sockets for name resolution. the `Client.resolve_http` test succeeds, but `Client.resolve_https` hangs after sending the request. still trying to debug that, but it doesn't look like libcurl ever asks us to wait for the response

I noticed this as well, and had to add CURLOPT_CONNECTTIMEOUT to see a connection timeout be used.

cbodley committed:
rgw/curl: move verbose logging to unit test

because async_perform() takes an existing CURL easy handle, it leaves
customization of that handle up to the caller. the only curl options
that Client changes on this easy handle are those necessary for
integration with the curl-multi api

while verbose logging is very useful, it should be managed outside of
this Client which doesn't otherwise have any dependency on CephContext
or its logging infrastructure

move this debug_callback() into the unit test as an example. other uses
in rgw adopt similar logging through dout

this also allows the unit test to avoid linkage to ceph global and
unit-main libraries

chardan commented on src/rgw/curl/async_perform.h:
```diff
@@ -0,0 +1,91 @@
+namespace detail {
+
+struct async_perform_impl {
+  void* easy = nullptr;
```
I sorta prefer easy_handle or easy_ptr, although your comment makes it perfectly understandable, so no need to change it.
